tsfc/driver.py

```diff
@@ -432,20 +432,51 @@ def compile_expression_dual_evaluation(expression, dual_functionals, coordinates
                      argument_multiindices=argument_multiindices)
 
     dual_expressions = []
-    for dual in dual_functionals:
-        # Extract associated quadrature rule representation
-        # Make a quad_rule
 
-        quad_points = [*dual.pt_dict.keys()]
-        # get it working for scalar functions first; how do we deal with components??
-        weights = [dual.pt_dict[pt][0][0] for pt in quad_points]
-        quad_rule = QuadratureRule(PointSet(quad_points), weights)
-
-        config = kernel_cfg.copy()
-        config.update(quadrature_rule=quad_rule)
-        expressions = [gem.index_sum(e, quad_rule.point_set.indices)
-                       for e in fem.compile_ufl(expression, **config)]
-        dual_expressions.append(expressions)
+    # We need to do different things based on the shape of the expression, unfortunately
+
+    # Case 1: scalars
+    if len(expression.ufl_shape) == 0:
+        for dual in dual_functionals:
+            quad_points = [*dual.pt_dict.keys()]
+            weights = [dual.pt_dict[pt][0][0] for pt in quad_points]
+            quad_rule = QuadratureRule(PointSet(quad_points), weights)
+
```

Comment:
See comment from first commit -- we need to assert that there is no deriv_dict or else this will fail silently.  Fixing this (and/or FIAT) to work with a deriv_dict is also a possibility.